src/shogun/kernel/MaternKernel.cpp

```diff
@@ -0,0 +1,197 @@
+	{
+		float64_t ratio = (std::sqrt(3) * dist) / m_width;
+		result = (1 + ratio) * std::exp(-ratio);
+	}
+	else if (Math::fequals(
+	             m_nu, 2.5, std::numeric_limits<float64_t>::epsilon()))
+	{
+		float64_t ratio = (std::sqrt(5) * dist) / m_width;
+		result = (1 + ratio +
+		          ((5 * std::pow(dist, 2)) / (3 * std::pow(m_width, 2)))) *
+		         std::exp(-ratio);
+	}
+	// if none of the above was true we compute the value the expensive way
+	// in theory if nu is large we could calculate this with the squared
+	// exponential covariance function
+	else
+	{
+		const float64_t adjusted_dist =
+		    Math::fequals(dist, 0.0, std::numeric_limits<float64_t>::epsilon())
+		        ? std::numeric_limits<float32_t>::epsilon()
+		        : dist;
+		float64_t ratio = std::sqrt(2 * m_nu) * adjusted_dist / m_width;
+		// bessel function of the second kind
+#ifdef __clang__
+		error(
+		    "Shogun compiled on clang does not support Matern kernels with "
+		    "width other than 0.5, 1.5 and 2.5. The current width is {}",
+		    m_width);
+#else
+		float64_t bessel = std::cyl_bessel_k(m_nu, ratio);
+#endif
+		result = (std::pow(2, 1 - m_nu) / std::tgamma(m_nu)) *
+		         std::pow(ratio, m_nu) * bessel;
+	}
+
+	return result;
+}
+
+SGMatrix<float64_t> MaternKernel::get_parameter_gradient(
+    Parameters::const_reference param, index_t index)
+{
+	require(lhs, "Left hand side features must be set!");
+	require(rhs, "Right hand side features must be set!");
+
+	if (param.first == "width")
+	{
+		SGMatrix<float64_t> derivative = SGMatrix<float64_t>(num_lhs, num_rhs);
+		std::function<float64_t(float64_t)> gradient_func;
+
+		// the gradients Matern wrt m_width were computed with WolframAlpha
+		if (Math::fequals(m_nu, 0.5, std::numeric_limits<float64_t>::epsilon()))
+		{
+			const float64_t m_width_squared = std::pow(m_width, 2);
+			gradient_func = [&](const float64_t& dist) {
+				return (dist * std::exp(-dist / m_width)) / m_width_squared;
+			};
+		}
+		else if (Math::fequals(
+		             m_nu, 1.5, std::numeric_limits<float64_t>::epsilon()))
+		{
+			const float64_t m_width_cubed = std::pow(m_width, 3);
+			gradient_func = [&](const float64_t& dist) {
+				return (3 * std::pow(dist, 2) *
+				        std::exp(-(std::sqrt(3) * dist) / m_width)) /
+				       m_width_cubed;
+			};
+		}
+		else if (Math::fequals(
+		             m_nu, 2.5, std::numeric_limits<float64_t>::epsilon()))
+		{
+			const float64_t m_width_power_4 = std::pow(m_width, 4);
+			gradient_func = [&](const float64_t& dist) {
+				return (5 * std::pow(dist, 2) *
+				        std::exp(-(std::sqrt(5) * dist) / m_width) *
+				        (m_width + std::sqrt(5) * dist)) /
+				       (3 * m_width_power_4);
+			};
+		}
+		else
+		{
+#ifdef __clang__
+			error(
+			    "Shogun compiled on clang does not support Matern kernels with "
+			    "width other than 0.5, 1.5 and 2.5. The current width is {}",
+			    m_width);
+#endif
```

Comment:
@karlnapf I couldn't find the bessel function I need here (`kv` in scipy) so I think we should just wait until clang implements it. It should also be possible to compile this with clang with libstdc++, rather than libc++

Reply:
problem will be tests / examples when compiled without clang, but ok

Reply:
I guess, but the option is adding some fortran code from http://netlib.org/amos/

Reply:
I can try and find out how stdlibc++ does it actually

Reply:
good idea. Not so sure about Fortran code, never did that. Is it easy?

Reply:
I copied the libstdc++ code :D